src/core/thread/mle_router.cpp

```diff
@@ -3327,6 +3329,8 @@ void MleRouter::RemoveNeighbor(Neighbor &aNeighbor)
     }
     else if (!IsActiveRouter(aNeighbor.GetRloc16()))
     {
+        OT_ASSERT(mChildTable.GetChildIndex(static_cast<Child &>(aNeighbor)) < kMaxChildren);
```

Comment:
Make sure the neighbor is a valid child.

Reply:
What situation did you have in mind here for this check (and the one for router)?

I think when we have `aNeighbor` pointer entry we have found/acquired it from either child table or router table.  Either case I think, it should be already in the related table with a valid index.

Reply:
We found two conditions that can incorrectly enter these two code blocks in the simulation. 

- `HandleAdvertisement` can incorrectly call `RemoveNeighbor` with `Router` from `mRouterTable`, which has `RLOC16 == 0xFFFF`. `RemoveNeighbor` will incorrectly handle this `aNeighbor` as a `Child`. 
- `HandleAdvertisement` can also call `RemoveNeighor` with `aNeighbor` being `mParentCandidate`. The original `RemoveNeighor` will incorrectly handle `aNeighbor` as `Child` or `Router` based on its `RLOC16`. 

In case there are more undiscovered bugs, I add these two `OT_ASSERT` statements for catching code bugs earlier.

Reply:
> Since aNeighbor is reset to NULL now, this scenario would not occur any more, right?

Yes, it should not happen any more. But I think we need to keep the `OT_ASSERT` because the very purpose of using `OT_ASSERT` is to catch something that **would not occur**.

Reply:
> How could the Router receive an MLE advertisement from the REED who has already changed its RLOC in such short 1.25s considering the time consumption of the attach better process of REED itself and the at least 1s jitter

Could it be that the REED was a Router and had just downgraded from being a Router? 

However without comprehensive debugging, we can't not know for sure why it happened.